services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceController.java

```diff
@@ -128,7 +119,10 @@ public VeniceController(VeniceControllerContext ctx) {
     this.routerClientConfig = Optional.ofNullable(ctx.getRouterClientConfig());
     this.icProvider = Optional.ofNullable(ctx.getIcProvider());
     this.externalSupersetSchemaGenerator = Optional.ofNullable(ctx.getExternalSupersetSchemaGenerator());
-    this.pubSubClientsFactory = Objects.requireNonNull(ctx.getPubSubClientsFactory(), "PubSubClientsFactory is null");
+    this.pubSubClientsFactory = new PubSubClientsFactory(
+        multiClusterConfigs.getPubSubProducerAdapterFactory(),
+        multiClusterConfigs.getPubSubConsumerAdapterFactory(),
+        multiClusterConfigs.getPubSubAdminAdapterFactory());
     createServices();
```

Comment:
I think we should just expose `PubSubClientsFactory getPubSubClientsFactory()` in `VeniceControllerMultiClusterConfig` instead of constructing `PubSubClientsFactory` here. Same for server too.